language/move-prover/abigen/src/abigen.rs

```diff
@@ -0,0 +1,190 @@
+            .map(|ty_param| {
+                TypeArgumentABI::new(symbol_pool.string(ty_param.0).to_string().to_lowercase())
+            })
+            .collect();
+        let args = func
+            .get_parameters()
+            .iter()
+            .filter_map(
+                |param| match self.get_type_tag_skipping_references(&param.1) {
+                    Ok(Some(tag)) => Some(Ok(ArgumentABI::new(
+                        symbol_pool.string(param.0).to_string(),
+                        tag,
+                    ))),
+                    Ok(None) => None,
+                    Err(error) => Some(Err(error)),
+                },
+            )
+            .collect::<anyhow::Result<_>>()?;
+        Ok(ScriptABI::new(name, doc, code, ty_args, args))
+    }
+
+    fn load_compiled_bytes(&self, module_env: &ModuleEnv<'env>) -> anyhow::Result<Vec<u8>> {
+        let mut path = PathBuf::from(&self.options.compiled_script_directory);
+        path.push(
+            PathBuf::from(module_env.get_source_path())
+                .with_extension("mv")
+                .file_name()
+                .expect("file name"),
+        );
+        let mut f = match std::fs::File::open(path.clone()) {
+            Ok(f) => f,
+            Err(error) => bail!("Failed to open compiled file {:?}: {}", path, error),
+        };
+        let mut bytes = Vec::new();
+        f.read_to_end(&mut bytes)?;
+        Ok(bytes)
+    }
+
+    fn get_type_tag_skipping_references(&self, ty0: &ty::Type) -> anyhow::Result<Option<TypeTag>> {
+        use ty::Type::*;
+        let tag = match ty0 {
+            Primitive(prim) => {
+                use ty::PrimitiveType::*;
+                match prim {
+                    Bool => TypeTag::Bool,
+                    U8 => TypeTag::U8,
+                    U64 => TypeTag::U64,
+                    U128 => TypeTag::U128,
+                    Address => TypeTag::Address,
+                    Signer => TypeTag::Signer,
+                    Num | Range | TypeValue => bail!("Type {:?} is not allowed in scripts.", ty0),
+                }
+            }
+            Reference(_, _) => {
+                // Skip references (most likely a `&signer` type)
+                return Ok(None);
+            }
+            Vector(ty) => {
+                let tag = self.get_type_tag(ty)?;
+                TypeTag::Vector(Box::new(tag))
+            }
+            Tuple(_)
+            | Struct(_, _, _)
+            | TypeParameter(_)
+            | Fun(_, _)
+            | TypeDomain(_)
+            | TypeLocal(_)
+            | Error
+            | Var(_) => bail!("Type {:?} is not allowed in scripts.", ty0),
+        };
+        Ok(Some(tag))
+    }
+
+    fn get_type_tag(&self, ty: &ty::Type) -> anyhow::Result<TypeTag> {
+        if let Some(tag) = self.get_type_tag_skipping_references(ty)? {
+            return Ok(tag);
+        }
+        bail!(
+            "References such as {:?} are only not allowed in the list of parameters.",
```

Comment:
Nit: Typo? I don't quite follow the meaning of "only not allowed".

Reply:
yes, it should be "only allowed"